### PR TITLE
Systems and UI

### LINKED ISSUE #87: Update balls position

### LINKED ISSUE #88: Balls rendering

### LINKED ISSUE #92: Create demo UI

### PR DESCRIPTION
**Pull request by vitlinda**
This PR introduces the `RenderingSystem`, the `MovementSystem` and the `FrictionSystem` along with the demo UI.
Closes #92 
Closes #88 
Closes #87

### CODE REVIEW

File: demo/src/main/scala/dev/atedeg/ecscalademo/SimulationStatus.scala [hunk 1]
```diff
@@ -9,5 +11,42 @@ object MouseState {
...
+    Velocity(Vector(0, 0)),
+    Velocity(Vector(0, 0)),
+    Velocity(Vector(0, 0)),
+    Velocity(Vector(0, 0)),
+    Velocity(Vector(0, 0)),
+  )
+
+  val radius = 20
+  val frictionCoefficient = 0.05
+  val gravity = 9.81
```
Comment (ndido98) [inline demo/src/main/scala/dev/atedeg/ecscalademo/SimulationStatus.scala:51]: In #99 they are in `EnvironmentState.scala`, where should we put them? @nicolasfara @giacomocavalieri

File: demo/src/main/scala/dev/atedeg/ecscalademo/controller/MainViewController.scala [hunk 2]
```diff
@@ -46,22 +48,26 @@ class MainViewController extends Initializable {
...
 
   override def initialize(url: URL, resourceBundle: ResourceBundle): Unit = {
-    loop = GameLoop(f => {
-      world.update(f.toFloat)
-    })
+    import dev.atedeg.ecscalademo.StartingState.*
+    val world = World()
+    for {
+      ((position, color), velocity) <- startingPositions zip startingColors zip startingVelocities
+    } world hasAn entity withComponents { Circle(radius, color) &: position &: velocity }
```
Comment (ndido98) [inline demo/src/main/scala/dev/atedeg/ecscalademo/controller/MainViewController.scala:60]: You're missing the `Mass` component

File: demo/src/test/scala/dev/atedeg/ecscalademo/systems/FrictionSystemTest.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        val initialVelocity = Vector(300, 0)
+        val ball = world hasAn entity withComponents {
+          Velocity(initialVelocity)
+        }
+        world hasA system(frictionSystem)
+
+        (0 to 2) foreach { _ => world.update(10) }
+
+        getView[Velocity &: CNil] from world should contain theSameElementsAs List(
+          (ball, Velocity(Vector(298.5285, 0)) &: CNil),
```
Comment (ndido98) [inline demo/src/test/scala/dev/atedeg/ecscalademo/systems/FrictionSystemTest.scala:29]: You can just check that the velocity has decreased if you prefer, so you can make a simpler test and remove that magic number

**Review by ndido98 — CHANGES_REQUESTED** [review]

File: demo/src/main/scala/dev/atedeg/ecscalademo/ECSCanvas.scala [hunk 4]
```diff
@@ -1,40 +1,73 @@
...
+   * @param color
+   *   The ball color.
+   * @param lineWidth
+   *   The thickness of the ball's border.
+   */
+  def drawCircle(coordinates: Point, radius: Double, color: Color, lineWidth: Double): Unit
+
+  /**
+   * @param from
+   *   The strating point from where draw the line.
```
Comment (giacomocavalieri) [inline demo/src/main/scala/dev/atedeg/ecscalademo/ECSCanvas.scala:28]:
```suggestion
   *   The starting point of the line.
```

File: demo/src/main/scala/dev/atedeg/ecscalademo/ECSCanvas.scala [hunk 5]
```diff
@@ -1,40 +1,73 @@
...
+   * @param to
+   *   The end point up to where draw the line.
```
Comment (giacomocavalieri) [inline demo/src/main/scala/dev/atedeg/ecscalademo/ECSCanvas.scala:30]:
```suggestion
   *   The end point of the line.
```

File: demo/src/main/scala/dev/atedeg/ecscalademo/systems/MovementSystem.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class MovementSystem extends System[Position &: Velocity &: CNil] {
+  override def shouldRun: Boolean = PlayState.playing
+
+  override def update(entity: Entity, components: Position &: Velocity &: CNil)(
+      deltaTime: DeltaTime,
+      world: World,
+      view: View[Position &: Velocity &: CNil],
+  ): Deletable[Position &: Velocity &: CNil] = {
+    val Position(point) &: Velocity(vector) &: CNil = components
```
Comment (giacomocavalieri) [inline demo/src/main/scala/dev/atedeg/ecscalademo/systems/MovementSystem.scala:20]: To me it read better like this
```suggestion
    val Position(position) &: Velocity(velocity) &: CNil = components
```

File: demo/src/main/scala/dev/atedeg/ecscalademo/systems/RenderSystem.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  override def before(deltaTime: DeltaTime, world: World, view: View[Circle &: Position &: CNil]): Unit =
+    ecsCanvas.clear()
+
+  override def update(entity: Entity, components: Circle &: Position &: CNil)(
+      deltaTime: DeltaTime,
+      world: World,
+      view: View[Circle &: Position &: CNil],
+  ): Deletable[Circle &: Position &: CNil] = {
+    val lineWidth = PlayState.selectedBall match { case Some(`entity`) => 3; case _ => 1 }
```
Comment (giacomocavalieri) [inline demo/src/main/scala/dev/atedeg/ecscalademo/systems/RenderSystem.scala:22]: This are magic numbers, what is 3?  What is 1? They could become variables named `selectedBallLineWidth` and `regularBallLineWidth`

**Review by giacomocavalieri — CHANGES_REQUESTED** [review]

File: demo/src/main/scala/dev/atedeg/ecscalademo/SimulationStatus.scala [hunk 8]
```diff
@@ -9,5 +11,45 @@ object MouseState {
...
+object StartingState {
+
+  val startingPositions = List(
+    Position(Point(147, 157)),
+    Position(Point(511, 157)),
+    Position(Point(546, 177)),
+    Position(Point(546, 136)),
+    Position(Point(581, 157)),
+    Position(Point(581, 117)),
+    Position(Point(581, 198)),
```
[the remark this answers is not in the document]
Comment (vitlinda) [reply demo/src/main/scala/dev/atedeg/ecscalademo/SimulationStatus.scala:26]: This is going to be done with the next PlayState implementation

File: demo/src/main/scala/dev/atedeg/ecscalademo/systems/FrictionSystem.scala [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  override def update(
+      entity: Entity,
+      components: Velocity &: CNil,
+  )(deltaTime: DeltaTime, world: World, view: View[Velocity &: CNil]): Deletable[Velocity &: CNil] = {
+    val Velocity(velocity) &: CNil = components
+    if (velocity.norm > 0) {
+      val frictionDirection = -1 * velocity.normalized
+      val friction = (frictionCoefficient * gravity) * frictionDirection
+      val newVelocity = velocity + friction
+      if ((velocity dot newVelocity) < 0) Velocity(Vector(0, 0)) &: CNil else Velocity(newVelocity) &: CNil
```
[the remark this answers is not in the document]
Comment (vitlinda) [reply demo/src/main/scala/dev/atedeg/ecscalademo/systems/FrictionSystem.scala:27]: Your are right, according to [https://docs.scala-lang.org/scala3/book/taste-control-structures.html](url)

**Review by nicolasfara — APPROVED** [review]

**Review by ndido98 — APPROVED** [review]

**Review by giacomocavalieri — APPROVED** [review]

**Pull request merged**